File: 16-permutation-promenade.jl
function eval_move!(move, dancers)
    move_type = move[1]
    params = move[2:end]
    if move_type == 's' # spin
        eval_spin!(params, dancers)
    elseif move_type == 'x' # exchange
        eval_exchange!(params, dancers)
    elseif move_type == 'p' # partner swap
        eval_partner!(params, dancers)
    end
end

function eval_spin!(params, dancers)
    shift = parse(Int, params)
    dancers[1:end] = circshift(dancers, shift)
end

function eval_exchange!(params, dancers)
    i, j = map(x -> parse(Int, x) + 1, split(params, "/"))
    dancers[i], dancers[j] = dancers[j], dancers[i]
end

function eval_partner!(params, dancers)
    a, b = split(params, "/")
    ia = findfirst([x == a for x in dancers])
    ib = findfirst([x == b for x in dancers])
    dancers[ia] = b
    dancers[ib] = a
end

function dance!(moves, dancers)
    for m in moves
        eval_move!(m, dancers)
    end
end

moves = split(readchomp(STDIN), ",")
dancers = collect(join(c) for c in 'a':'p')
orig_dancers = copy(dancers)
dance_cycle = [orig_dancers]

dance!(moves, dancers)
println(join(dancers))

while dancers != orig_dancers
    push!(dance_cycle, copy(dancers))
    dance!(moves, dancers)
end

println(join(dance_cycle[1_000_000_000 % length(dance_cycle) + 1]))

File: 16-permutation-promenade.py
from collections import deque

parse_params = {
    's': lambda x: int(x),
    'x': lambda x: tuple(map(int, x.split('/'))),
    'p': lambda x: tuple(x.split('/')),
}


def parse_move(move):
    move_type = move[0]
    return (move_type, parse_params[move_type](move[1:]))


def eval_spin(shift, dancers):
    dancers.rotate(shift)


def eval_exchange(params, dancers):
    i, j = params
    dancers[i], dancers[j] = dancers[j], dancers[i]


def eval_partner(params, dancers):
    a, b = params
    i = dancers.index(a)
    j = dancers.index(b)
    dancers[i], dancers[j] = dancers[j], dancers[i]


evaluators = dict(s=eval_spin, x=eval_exchange, p=eval_partner)


moves = list(map(parse_move, input().split(',')))
dancers = deque("abcdefghijklmnop")
orig_dancers = dancers.copy()
cycle = [tuple(dancers)]

for m, p in moves:
    evaluators[m](p, dancers)
print(''.join(dancers))

while dancers != orig_dancers:
    cycle.append(tuple(dancers))
    for m, p in moves:
        evaluators[m](p, dancers)

print(''.join(cycle[1_000_000_000 % len(cycle)]))
